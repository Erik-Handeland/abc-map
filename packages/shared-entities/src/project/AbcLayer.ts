import { FeatureCollection } from 'geojson';
import { AbcProjection } from './AbcProjection';

export type AbcLayer = AbcVectorLayer | AbcPredefinedLayer | AbcWmsLayer;

export type LayerMetadata = VectorMetadata | PredefinedMetadata | WmsMetadata;

export interface AbcBaseLayer {
  type: LayerType;
  metadata: LayerMetadata;
}

export interface BaseMetadata {
  id: string;
  name: string;
  opacity: number;
  visible: boolean;
  active: boolean;
  type: LayerType;
}

export interface VectorMetadata extends BaseMetadata {
  type: LayerType.Vector;
}

export interface PredefinedMetadata extends BaseMetadata {
  type: LayerType.Predefined;
  model: PredefinedLayerModel;
}

export interface WmsMetadata extends BaseMetadata {
  type: LayerType.Wms;
  url: string;
  layerName: string;
  projection?: AbcProjection;
  extent?: [number, number, number, number];
  auth?: {
    username: string;
    password: string;
  };
}

export enum LayerType {
  Vector = 'Vector',
  Predefined = 'Predefined',
  Wms = 'Wms',
}

export interface AbcVectorLayer extends AbcBaseLayer {
  type: LayerType.Vector;
  metadata: VectorMetadata;
  features: FeatureCollection;
}

export interface AbcPredefinedLayer extends AbcBaseLayer {
  type: LayerType.Predefined;
  metadata: PredefinedMetadata;
}

export interface AbcWmsLayer extends AbcBaseLayer {
  type: LayerType.Wms;
  metadata: WmsMetadata;
}

export enum PredefinedLayerModel {
  OSM = 'OSM',
}
