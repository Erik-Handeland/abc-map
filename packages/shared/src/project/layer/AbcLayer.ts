/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { FeatureCollection } from 'geojson';
import { AbcProjection } from '../AbcProjection';

export type AbcLayer = AbcVectorLayer | AbcPredefinedLayer | AbcWmsLayer;

export type LayerMetadata = VectorMetadata | PredefinedMetadata | WmsMetadata;

export interface AbcBaseLayer {
  type: LayerType;
  metadata: LayerMetadata;
}

export interface BaseMetadata {
  id: string;
  name: string;
  opacity: number;
  visible: boolean;
  active: boolean;
  type: LayerType;
}

export interface VectorMetadata extends BaseMetadata {
  type: LayerType.Vector;
}

export interface PredefinedMetadata extends BaseMetadata {
  type: LayerType.Predefined;
  model: PredefinedLayerModel;
}

export interface WmsMetadata extends BaseMetadata {
  type: LayerType.Wms;
  projection?: AbcProjection;
  extent?: [number, number, number, number];
  remoteUrl: string;
  remoteLayerName: string;
  auth?: {
    username: string;
    password: string;
  };
}

export enum LayerType {
  Vector = 'Vector',
  Predefined = 'Predefined',
  Wms = 'Wms',
}

export interface AbcVectorLayer extends AbcBaseLayer {
  type: LayerType.Vector;
  metadata: VectorMetadata;
  features: FeatureCollection;
}

export interface AbcPredefinedLayer extends AbcBaseLayer {
  type: LayerType.Predefined;
  metadata: PredefinedMetadata;
}

export interface AbcWmsLayer extends AbcBaseLayer {
  type: LayerType.Wms;
  metadata: WmsMetadata;
}

export enum PredefinedLayerModel {
  OSM = 'OSM',
  StamenToner = 'StamenToner',
  StamenTonerLite = 'StamenTonerLite',
  StamenTerrain = 'StamenTerrain',
  StamenWatercolor = 'StamenWatercolor',
}
