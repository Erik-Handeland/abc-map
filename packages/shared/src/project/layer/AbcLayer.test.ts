/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcPredefinedLayer, AbcVectorLayer, AbcWmsLayer, AbcXyzLayer, LayerType, PredefinedLayerModel } from './AbcLayer';

/**
 * If this test fail, you should write a migration script then adapt it
 */
describe('AbcLayer', () => {
  it('VectorLayer should not change without migration', () => {
    /* eslint-disable */
    const witness = '{"type":"Vector","metadata":{"id":"test-layer-id","name":"Test vector layer","active":true,"visible":true,"opacity":0.5,"type":"Vector"},"features":{"type":"FeatureCollection","bbox":[1,2,3,4],"features":[{"id":"feature-id","type":"Feature","bbox":[5,6,7,8],"properties":{"property1":"value1"},"geometry":{"type":"Point","coordinates":[9,10]}}]}}';
    /* eslint-enable */

    const current: AbcVectorLayer = {
      type: LayerType.Vector,
      metadata: {
        id: 'test-layer-id',
        name: 'Test vector layer',
        active: true,
        visible: true,
        opacity: 0.5,
        type: LayerType.Vector,
      },
      features: {
        type: 'FeatureCollection',
        bbox: [1, 2, 3, 4],
        features: [
          {
            id: 'feature-id',
            type: 'Feature',
            bbox: [5, 6, 7, 8],
            properties: {
              property1: 'value1',
            },
            geometry: {
              type: 'Point',
              coordinates: [9, 10],
            },
          },
        ],
      },
    };

    expect(JSON.stringify(current)).toEqual(witness);
  });

  it('PredefinedLayer should not change without migration', () => {
    /* eslint-disable */
    const layerWitness = '{"type":"Predefined","metadata":{"id":"test-layer-id","name":"Test predefined layer","active":true,"visible":true,"opacity":0.5,"type":"Predefined","model":"OSM"}}';
    const modelsWitness = '"{\\"OSM\\":\\"OSM\\",\\"StamenToner\\":\\"StamenToner\\",\\"StamenTonerLite\\":\\"StamenTonerLite\\",\\"StamenTerrain\\":\\"StamenTerrain\\",\\"StamenWatercolor\\":\\"StamenWatercolor\\"}"';
    /* eslint-enable */

    const currentLayer: AbcPredefinedLayer = {
      type: LayerType.Predefined,
      metadata: {
        id: 'test-layer-id',
        name: 'Test predefined layer',
        active: true,
        visible: true,
        opacity: 0.5,
        type: LayerType.Predefined,
        model: PredefinedLayerModel.OSM,
      },
    };
    const currentModels = JSON.stringify(PredefinedLayerModel);

    expect(JSON.stringify(currentLayer)).toEqual(layerWitness);
    expect(JSON.stringify(currentModels)).toEqual(modelsWitness);
  });

  it('WMS layer should not change without migration', () => {
    /* eslint-disable */
    const layerWitness = '{"type":"Wms","metadata":{"id":"test-layer-id","name":"Test wms layer","active":true,"visible":true,"opacity":0.5,"type":"Wms","auth":{"username":"test-username","password":"test-password"},"extent":[1,1,1,1],"projection":{"name":"EPSG:4326"},"remoteLayerName":"test-remoteLayerName","remoteUrls":["test-remoteUrls"]}}';
    /* eslint-enable */

    const currentLayer: AbcWmsLayer = {
      type: LayerType.Wms,
      metadata: {
        id: 'test-layer-id',
        name: 'Test wms layer',
        active: true,
        visible: true,
        opacity: 0.5,
        type: LayerType.Wms,
        auth: {
          username: 'test-username',
          password: 'test-password',
        },
        extent: [1, 1, 1, 1],
        projection: { name: 'EPSG:4326' },
        remoteLayerName: 'test-remoteLayerName',
        remoteUrls: ['test-remoteUrls'],
      },
    };

    expect(JSON.stringify(currentLayer)).toEqual(layerWitness);
  });

  it('XYZ layer should not change without migration', () => {
    /* eslint-disable */
    const layerWitness = '{"type":"Xyz","metadata":{"id":"test-layer-id","name":"Test wms layer","active":true,"visible":true,"opacity":0.5,"type":"Xyz","remoteUrl":"test-remoteUrl"}}';
    /* eslint-enable */

    const currentLayer: AbcXyzLayer = {
      type: LayerType.Xyz,
      metadata: {
        id: 'test-layer-id',
        name: 'Test wms layer',
        active: true,
        visible: true,
        opacity: 0.5,
        type: LayerType.Xyz,
        remoteUrl: 'test-remoteUrl',
      },
    };

    expect(JSON.stringify(currentLayer)).toEqual(layerWitness);
  });
});
