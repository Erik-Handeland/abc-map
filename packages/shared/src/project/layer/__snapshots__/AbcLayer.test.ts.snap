// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbcLayer PredefinedLayer should not change without migration 1`] = `
{
  "metadata": {
    "active": true,
    "id": "test-layer-id",
    "model": "OSM",
    "name": "Test predefined layer",
    "opacity": 0.5,
    "type": "Predefined",
    "visible": true,
  },
  "type": "Predefined",
}
`;

exports[`AbcLayer PredefinedLayer should not change without migration 2`] = `
{
  "OSM": "OSM",
  "StamenTerrain": "StamenTerrain",
  "StamenToner": "StamenToner",
  "StamenTonerLite": "StamenTonerLite",
  "StamenWatercolor": "StamenWatercolor",
}
`;

exports[`AbcLayer VectorLayer should not change without migration 1`] = `
{
  "features": {
    "bbox": [
      1,
      2,
      3,
      4,
    ],
    "features": [
      {
        "bbox": [
          5,
          6,
          7,
          8,
        ],
        "geometry": {
          "coordinates": [
            9,
            10,
          ],
          "type": "Point",
        },
        "id": "feature-id",
        "properties": {
          "property1": "value1",
        },
        "type": "Feature",
      },
    ],
    "type": "FeatureCollection",
  },
  "metadata": {
    "active": true,
    "id": "test-layer-id",
    "name": "Test vector layer",
    "opacity": 0.5,
    "type": "Vector",
    "visible": true,
  },
  "type": "Vector",
}
`;

exports[`AbcLayer WMS layer should not change without migration 1`] = `
{
  "metadata": {
    "active": true,
    "auth": {
      "password": "test-password",
      "username": "test-username",
    },
    "extent": [
      1,
      1,
      1,
      1,
    ],
    "id": "test-layer-id",
    "name": "Test wms layer",
    "opacity": 0.5,
    "projection": {
      "name": "EPSG:4326",
    },
    "remoteLayerName": "test-remoteLayerName",
    "remoteUrls": [
      "test-remoteUrls",
    ],
    "type": "Wms",
    "visible": true,
  },
  "type": "Wms",
}
`;

exports[`AbcLayer XYZ layer should not change without migration 1`] = `
{
  "metadata": {
    "active": true,
    "id": "test-layer-id",
    "name": "Test wms layer",
    "opacity": 0.5,
    "remoteUrl": "test-remoteUrl",
    "type": "Xyz",
    "visible": true,
  },
  "type": "Xyz",
}
`;
